File: app/src/main/java/at/yawk/fiction/android/download/ProgressListener.java
package at.yawk.fiction.android.download;

/**
 * @author yawkat
 */
public interface ProgressListener {
    ProgressListener NOOP = new ProgressListener() {
        @Override
        public ProgressListener createSubLevel() {
            return NOOP;
        }

        @Override
        public void progressDeterminate(long progress, long limit) {}

        @Override
        public void progressIndeterminate(boolean complete) {}
    };

    /**
     * Create a new progress sub-level. Note that multiple sub-levels may exist and be used concurrently.
     */
    ProgressListener createSubLevel();

    /**
     * Publish a new progress report with a known upper bound.
     */
    void progressDeterminate(long progress, long limit);

    /**
     * Publish a new progress report without known progress indication.
     *
     * @param complete <code>true</code> if this task was completed.
     */
    void progressIndeterminate(boolean complete);
}

File: app/src/main/kotlin/at/yawk/fiction/android/download/ProgressListener.kt
package at.yawk.fiction.android.download

/**
 * @author yawkat
 */
interface ProgressListener {
    fun postProgress(progress: Progress)

    interface Progress

    data class Determinate(
            val message: String,
            val complete: Int,
            val total: Int
    ) : Progress

    data class Indeterminate(val message: String) : Progress

    object Complete : Progress
}